cypress-io/cypress
# fix: Properly type Cypress and Chai globals as properties of the global object

Minor modifications to CLI types so that Cypress and Chai globals are correctly declared not only to be in the global scope but also to be in properties of globalThis, which they are. The change allows composers to reference them through the global object (e.g., 'globalThis.cy.mount', 'globalThis.expect', 'window.assert'). The current typing prohibits this by incorrectly declaring that the objects in question are held in local variables of the global scope rather than in properties of the global object.



- Closes [Issue 27539](https://github.com/cypress-io/cypress/issues/27539)

### Additional details

- Why was this change necessary?
This small change accommodates a coding style convention that demands that values in properties of the global object be accessed via the globalThis property of that object to enhance clarity regarding the origin of the values as globals as opposed to values defined in and imported from modules.
- What is affected by this change?
The change has no impact on the TypeScript compiler's acceptance of code that uses Cypress globals without that convention, which is the majority of such code.
- Any implementation details to explain?
Resolving the issue entailed swapping the use of `const` for `var` in type declarations in cli/types/cypress-expect.d.ts and cli/types/cypress-global-vars.d.ts.

### Steps to test

Ensure that typical code that leverages Cypress like this is still is accepted by the TypeScript compiler:

```ts
const browserVersion = window.Cypress.browser.version;
window.cy.mount(AppComponent);
window.assert("foo" !== "bar", "foo is not bar");
window.expect(true).not.to.be.false;
```

Then confirm that the TypeScript compiler also issues no errors regarding code that follows the aforementioned convention like the following:

```ts
const browserVersion = window.Cypress.browser.version;
window.cy.mount(AppComponent);
window.assert("foo" !== "bar", "foo is not bar");
window.expect(true).not.to.be.false;
```

```ts
const browserVersion = globalThis.Cypress.browser.version;
globalThis.cy.mount(AppComponent);
globalThis.assert("foo" !== "bar", "foo is not bar");
globalThis.expect(true).not.to.be.false;
```

### How has the user experience changed?


Before:
<img width="673" alt="image" src="https://github.com/cypress-io/cypress/assets/62110041/2fbb2210-425f-4a48-8c1a-ca59e65a1a3b">

After:
<img width="686" alt="image" src="https://github.com/cypress-io/cypress/assets/62110041/3ffac3f0-8935-4e2b-b4f3-a3a5deec80f5">

### PR Tasks


- [na] Have tests been added/updated?
- [no] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? (Little to no value in drawing attention in the documentation to the fact that `globalThis.cy` and `window.cy` now work just as well as `cy` in TypeScript code.)
- [yes] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
